### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by notalfredo**
- **Please check if the PR fulfills these requirements**
   - [ ] The commit message follows our guidelines
   - [ ] Tests for the changes have been added (for bug fixes / features)
   - [ ] Docs have been added / updated (for bug fixes / features)
 - **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
 - **What is the current behavior?** (You can also link to an open issue here)
 - **What is the new behavior (if this is a feature change)?**
 - **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
 - **Other information**:

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1,6 +1,560 @@
...
+            StringDiffOpKind::Insert(x) => {
+                new_string.push(x);
+            },
+            StringDiffOpKind::Substitute(x,y) => {
+                new_string.replace_range(
+                    (i.index as usize)..( (i.index as usize)+1 as usize),
+                &y.to_string())
+            },
+            _ => {
+                panic!("Unrecoginzed Opperation")
```
Comment (neoncitylights) [inline src/lib.rs:51]: Typo, should be "Operation"

File: src/lib.rs [hunk 2]
```diff
@@ -1,6 +1,560 @@
...
+    Insert(char),
+    Delete(char),
+    Transpose(usize, usize),
+}
+
+#[derive(Debug)]
+pub struct StringDiffOp {
+    pub kind: StringDiffOpKind,
+    pub index: isize,
+}
```
Comment (neoncitylights) [inline src/lib.rs:18]: There should be a constructor for this type, which should look like:
```rs
impl StringDiffOp {
   pub fn new(kind: StringDiffOpKind, index: isize) {
       Self { kind, index }
   }
}
```

And any instance of creating a new `StringDiffOp` should use this, via `StringDiffOp::new()`

File: src/lib.rs [hunk 3]
```diff
@@ -1,6 +1,560 @@
...
+    Substitute(char, char),  
+    Insert(char),
+    Delete(char),
+    Transpose(usize, usize),
+}
+
+#[derive(Debug)]
+pub struct StringDiffOp {
+    pub kind: StringDiffOpKind,
+    pub index: isize,
```
Comment (neoncitylights) [inline src/lib.rs:17]: Is there a specific reason this is an `isize` (signed)? This would make more sense as a `usize` (unsigned) since you can't negatively index a string

File: src/lib.rs [hunk 4]
```diff
@@ -1,6 +1,560 @@
...
+}
+
+pub trait StringDiffAlgorithm {
+    //fn diff<'a>(s1: &'a str, s2: &'a str) -> Vec<StringDiffOp>;
+    fn diff<'a>(&self, s1: &'a str, s2: &'a str) -> Vec<StringDiffOp>;
+    fn distance<'a>(&self, s1: &'a str, s2: &'a str) -> usize;
+}
+
+pub struct HammingDistance {}
+impl HammingDistance{}
```
Comment (neoncitylights) [inline src/lib.rs:65]: Since this is an empty `impl`, we can remove this

Comment (neoncitylights) [inline src/lib.rs:65]: Since this `impl` block is empty, we can safely remove this

File: src/lib.rs (hunk 2 above)
Comment (neoncitylights) [inline src/lib.rs]: Could you also derive the `PartialEq` and `Eq` traits for StringDiffOp? This allows overloading the `==` and `!=` operators and comparing them against each other.

For the `index` property: since `isize` is signed, it doesn't really make sense because string indices are `usize` (unsigned, zero or positive), where as signed would imply possibly negative indices. So, this should be the `usize` type.

----

There should be a constructor for the `StringDiffOp` type, which would look like:
```rs
impl StringDiffOp {
   // a generic constructor
   pub fn new(kind: StringDiffOpKind, index: usize) -> Self {
       Self {
           kind,
           index
       }
   }
   
   // utility method to create a delete-specific StringDiffOp
   pub fn new_delete(c: char, index: usize) -> Self {
       Self::new(StringDiffOpKind::Delete(c), index)
   }
}
```

And then callers can create a new instance of `StringDiffOp` via `StringDiffOp::new()`, and pass in the appropriate arguments. Same with `StringDiffOp::new_delete()`, `StringDiffOp::new_insert()`, etc.

The utility constructors also make it easier since otherwise would be a bit verbose+repetitive to keep doing the same thing. Compare:

```rust
// original
StringDiffOp {
    kind: StringDiffOpKind::Delete('c')
    index: 1usize,
}

// second (but StringDiffOpKind::new() can still be public)
StringDiffOpKind::new(StringDiffOp::Delete('c'), 1usize)

// now
StringDiffOpKind::new_delete('c', 1usize)
```

File: src/lib.rs (hunk 1 above)
Comment (neoncitylights) [inline src/lib.rs]: There should be a match arm for `StringDiffOpKind::Transpose`, and since that's not currently not supported you can panic with a message similar to "applyDiff() does not currently support the transpose operation yet." ("Opperation" is also a typo). 

Since the `match` construct can do [exhaustive matching](https://doc.rust-lang.org/book/ch06-02-match.html#matches-are-exhaustive), you can get rid of the default match arm (`_`) after adding one for `Transpose`.

File: src/lib.rs [hunk 5]
```diff
@@ -1,6 +1,560 @@
...
+            _ => {
+                panic!("Unrecoginzed Opperation")
+            }
+        }
+    }
+    return new_string.clone();
+}
+
+pub trait StringDiffAlgorithm {
+    //fn diff<'a>(s1: &'a str, s2: &'a str) -> Vec<StringDiffOp>;
```
Comment (neoncitylights) [inline src/lib.rs]: This commented out line can be safely removed

File: src/lib.rs (hunk 4 above)
Comment (neoncitylights) [inline src/lib.rs]: Since this is an empty `impl` block, this can be safely removed

File: src/lib.rs [hunk 6]
```diff
@@ -1,6 +1,560 @@
...
+        println!("{:?}" , dist_vector);
+        return dist_vector[first_string_len][second_string_len];
+    }
+}
+
+
+#[cfg(test)]
+mod dcode_tests{
+
+    fn eq_with_nan_eq(a: &StringDiffOp, b: &StringDiffOp) -> bool {
```
Comment (neoncitylights) [inline src/lib.rs]: By deriving the equality-related traits for `StringDiffOp` (mentioned above), you could just do `b == a` and that'd work, and you can take out this function since it won't be necessary anymore

File: src/lib.rs [hunk 7]
```diff
@@ -1,6 +1,560 @@
...
+        });           
+
+
+        assert_eq!(vec_compare(&test_vec, &test_struct.diff("karolin", "kathrin")) , true );
+        assert_eq!(vec_compare(&test_vec_2, &test_struct.diff("karolin", "kerstin")) , true );
+        assert_eq!(vec_compare(&test_vec_3, &test_struct.diff("kathrin", "kerstin")) , true );
+        assert_eq!(vec_compare(&test_vec_4, &test_struct.diff("0000", "1111")) , true );
+        assert_eq!(vec_compare(&test_vec_5, &test_struct.diff("2173896", "2233796")) , true );
+
+        //assert_eq!( (test_vec_5[0].kind).0 ,super::StringDiffOpKind::Substitute('7', '3'));
```
Comment (neoncitylights) [inline src/lib.rs]: Why was this commented out?

File: src/lib.rs [hunk 8]
```diff
@@ -1,6 +1,560 @@
...
+}
+
+#[derive(Debug)]
+pub struct StringDiffOp {
+    pub kind: StringDiffOpKind,
+    pub index: isize,
+}
+
+pub (crate) fn remove(start: usize, stop: usize, s: &str) -> String {
+    let mut rslt = "".to_string();
```
Comment (neoncitylights) [inline src/lib.rs]: - At the beginning of the function, there should be a preliminary check to make sure stop is greater than start. This error checking will cause a panic for the caller if they don't pass in the correct arguments.
   ```rust
   assert!(stop > start);
   ```
 - The variable name should be `result` since it's more readable
 - We can a create a new empty `String` type immediately via `String::new()`. This is nice since it also doesn't require changing types; `"".to_string()` creates a `&str` and then converts it to `String`

File: src/lib.rs [hunk 9]
```diff
@@ -1,6 +1,560 @@
...
+                    opp_vec.push(new_opp)
+                }
+            }
+        }
+        return opp_vec;
+    }        
+
+    
+    fn distance<'a>(&self, s1: &'a str, s2: &'a str) -> usize{
+        let mut edit_distance = 0;
```
Comment (neoncitylights) [inline src/lib.rs]: This statement initializing the `edit_distance` variable should go after the error checking to avoid allocating memory whenever the caller hits that error. It should also specify the type via `0usize`

File: src/lib.rs [hunk 10]
```diff
@@ -1,6 +1,560 @@
...
+    pub(crate) fn min_dist(x: usize, y: usize, z: usize) -> usize {
+        if x <= y && x <= z{
+            return x;
+        }
+        else if y <= x && y <= z {
+            return y;
+        }
+        else{
+            return z;
+        }
```
Comment (neoncitylights) [inline src/lib.rs]: If I'm understanding this correctly, this tries to find the smallest number between three numbers, right? This can be simplified with two calls to `std::cmp::min()` ([docs](https://doc.rust-lang.org/std/cmp/fn.min.html)), like:

```rust
use std::cmp::min;

pub(crate) fn min_dist(x: usize, y: usize, z: usize) -> usize {
    min(x, min(y, z))
}
```

File: src/lib.rs [hunk 11]
```diff
@@ -1,6 +1,560 @@
...
+        assert_eq!(3 , test_struct.distance("2173896", "2233796"));
+    }
+
+    #[test] 
+    fn test_HammingDistance_opp_distance(){
+        let test_struct = super::HammingDistance{};
+
+        let mut test_vec: Vec<StringDiffOp> = Vec::new();
+        test_vec.push(
+                super::StringDiffOp{
```
Comment (neoncitylights) [inline src/lib.rs]: These calls can be made more succinct after creating the utility constructors, e.g this could be `StringDiffOp::new_substitute()`

File: src/lib.rs [hunk 12]
```diff
@@ -1,6 +1,560 @@
...
+        for i in my_vector.iter(){
+            println!("{:?}", i);
+        }
+    }
+
+    pub(crate) fn get_operations(my_opp: &Vec<Vec<char>>, 
+                                    left_string: &str,
+                                    top_string: &str) -> Vec<StringDiffOp> {
+                        
+        let mut my_opp_vector : Vec<StringDiffOp> = Vec::new();
```
Comment (neoncitylights) [inline src/lib.rs]: I think this would make more sense to be called `diff_ops` instead (operations also only has 1 p)

File: src/lib.rs [hunk 13]
```diff
@@ -1,6 +1,560 @@
...
+        loop{
+            
+            if top_str_len == 0 && left_str_len == 0{
+                break;
+            }
+                        //Rows               Columns
+            match my_opp[left_str_len][top_str_len] {
+                //insertion
+                '^' => {
+                    let insertionOpp : StringDiffOp = StringDiffOp{
```
Comment (neoncitylights) [inline src/lib.rs]: this should be `insertion_op` (variables are traditionally `camel_case`), also only 1 p, same for the instance below

File: src/lib.rs [hunk 14]
```diff
@@ -1,6 +1,560 @@
...
+
+                },
+                //substitution
+                '\\' => {
+                    if prev_char == '^'{
+                        my_opp_vector.reverse();
+                    }
+ 
+                    if left_string.chars().nth(left_str_len-1).unwrap() != top_string.chars().nth(top_str_len-1).unwrap(){
+                        let insertionOpp : StringDiffOp = StringDiffOp{
```
Comment (neoncitylights) [inline src/lib.rs]: Copy/paste error? :P This should be `substitute_op`

File: src/lib.rs [hunk 15]
```diff
@@ -1,6 +1,560 @@
...
+        for row in dir_vector.iter(){
+            println!("{:?}", row)
+        }
+        */
+
+        //println!("Made it here");
+        let test = LevenshteinDistance::get_operations(&dir_vector, s1, s2);
+        println!("{:?}", test);
+        //return dist_vector[first_string_len][second_string_len];
+        return test;
```
Comment (neoncitylights) [inline src/lib.rs]: There's some debug stuff leftover that should be taken out, and then immediately return `LevenshteinDistance::get_operations(&dir_vector, s1, s2)` without creating a new variable

File: src/lib.rs [hunk 16]
```diff
@@ -1,6 +1,560 @@
...
+        }
+        else if y <= x && y <= z {
+            return y;
+        }
+        else{
+            return z;
+        }
+    }
+
+    pub(crate) fn min_dist_with_dir(x: usize, y: usize, z: usize) -> (usize,char){
```
Comment (neoncitylights) [inline src/lib.rs]: Could you explain a little bit about what this function does? I'm assuming `dir` means direction here? I can see that down below each one represents a kind of diff op, but each character seems arbitrarily chosen and can be a little hard to follow until you look down below. This is also a good opportunity to write a docblock for the function with `///`

Maybe with some clarification we can make this a bit more readable! 👍

File: src/lib.rs [hunk 17]
```diff
@@ -1,6 +1,560 @@
...
+                    left_str_len -= 1;
+                    top_str_len -= 1; 
+                    prev_char = '\\';
+                },
+                //deletion
+                '<' => {
+                    if prev_char == '^'{
+                        my_opp_vector.reverse();
+                    }
+                    let insertionOpp : StringDiffOp = StringDiffOp{
```
Comment (neoncitylights) [inline src/lib.rs]: `delete_op`

**Review by neoncitylights — CHANGES_REQUESTED** [review]
Thank you so much for writing this! :D Overall, I really like the direction of this code. I wrote some initial comments for the biggest things that stick out to me.

Can you run `rustfmt` and `clippy` as well? `Rustfmt` is the code-formatting tool, and `clippy` acts as the linter (think of like ESLint and Prettifier)

File: src/lib.rs (hunk 7 above)
Comment (notalfredo) [reply src/lib.rs]: Ignore that junk ill delete it (-｡-;

File: src/lib.rs [hunk 18]
```diff
@@ -1,6 +1,458 @@
...
+	pub(crate) fn min_dist(x: usize, y: usize, z: usize) -> usize {
+		cmp::min(x, cmp::min(y, z))
+	}
+	/// At a given (x,y) we must choose the minimum value between a cells
+	/// Top, Left, and Diagnal value. Depending on which cell is chosen between
+	/// the three it will tell us if its a deletion, insertion or substitution operation.
+	/// if we chooze x(The value above the cell) as the min value its a insertion operation (symbolized by '^')
+	/// if we choose y(The value left of the cell) as the min value its a deletion operation(symbolized by '<')
+	/// if we choose z(The value diagnal of the cell) as the min value its a substitution operation( sybmolized by '\' )
+	/// else we get an x (should never happen since either(x,y,z should be the smalles))
```
Comment (neoncitylights) [inline src/lib.rs]: >else we get an x (should never happen since either(x,y,z should be the **smalles**))

typo, should be "smallest". Also, since that condition should never occur, then you can replace the line `(0, 'x')` with the `unreachable!()` macro ([docs](https://doc.rust-lang.org/std/macro.unreachable.html)).

File: src/lib.rs [hunk 19]
```diff
@@ -1,6 +1,455 @@
...
+		let mut prev_char: char = ' ';
+
+		loop {
+			if top_str_len == 0 && left_str_len == 0 && prev_char == '^' {
+				Self::reverse_vec_and_indexes(&mut diff_ops, top_str_len);
+				break;
+			}
+			if top_str_len == 0 && left_str_len == 0 {
+				break;
+			}
```
Comment (neoncitylights) [inline src/lib.rs]: This block of code can be a little bit more succinct, since there's a few conditions checked twice (checking the length of both strings).
```rs
if top_str_len == 0 && left_str_len == 0 {
	break;
}

if prev_char == '^' {
	Self::reverse_vec_and_indexes(&mut diff_ops, top_str_len);
	break;
}
```

File: src/lib.rs [hunk 20]
```diff
@@ -1,6 +1,455 @@
...
+	let mut result = String::from("");
+	for (i, c) in s.chars().enumerate() {
+		if start > i || stop < i + 1 {
+			result.push(c);
+		}
+	}
+	result
+}
+
+pub fn aplly_diff(s: &str, diffs: Vec<StringDiffOp>) -> String {
```
Comment (neoncitylights) [inline src/lib.rs]: Sorry I didn't catch this, this is a typo. It should be "apply" (not aplly)

File: src/lib.rs [hunk 21]
```diff
@@ -1,6 +1,455 @@
...
+				new_string = remove(i.index, i.index + 1, &new_string);
+			}
+			StringDiffOpKind::Insert(_x) => {
+				new_string.push(_x);
+			}
+			StringDiffOpKind::Substitute(_x, _y) => {
+				new_string.replace_range((i.index)..(i.index + 1), &_y.to_string())
+			}
+			StringDiffOpKind::Transpose(_x, _y) => {
+				panic!("aplly_diff does not currently support the transpose operation yet")
```
Comment (neoncitylights) [inline src/lib.rs]: `aplly_diff` -> `apply_diff()`

File: src/lib.rs [hunk 22]
```diff
@@ -1,6 +1,455 @@
...
+	}
+}
+
+pub struct LevenshteinDistance {}
+impl LevenshteinDistance {
+	pub(crate) fn min_dist(x: usize, y: usize, z: usize) -> usize {
+		cmp::min(x, cmp::min(y, z))
+	}
+	/// At a given (x,y) we must choose the minimum value between a cells
+	/// Top, Left, and Diagnal value. Depending on which cell is chosen between
```
Comment (neoncitylights) [inline src/lib.rs]: `Diagnal` -> `Diagonal`

**Review by neoncitylights — CHANGES_REQUESTED** [review]
This will be the last review for this set of code! Just minor changes, and I'll go ahead and merge this after. I'll open up some issues so we can follow-up! Thanks a bunch.

File: src/lib.rs [hunk 23]
```diff
@@ -1,6 +1,455 @@
...
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct StringDiffOp {
+	pub kind: StringDiffOpKind,
+	pub index: usize,
+}
+
+impl StringDiffOp {
+	// a generic constructor
```
Comment (neoncitylights) [inline src/lib.rs]: Sorry, I meant this comment as an explanation to you in the code review (to help explain the difference between the constructors); we can delete this comment
